plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/organize/imports/ImportsComputer.xtend

```diff
@@ -145,271 +148,216 @@ public class ImportsComputer {
-				val defImp = impSpec.filter(DefaultImportSpecifier).head; // only one is possible
-				val defaultImport = if (defImp === null) "" else '''«defImp.importedElement.name», ''';
-
-				'''import «defaultImport»{«spacer»«FOR a : impSpec SEPARATOR ', '»«(a as NamedImportSpecifier).importedElement.name»«
-					IF ((a as NamedImportSpecifier).alias !== null)» as « (a as NamedImportSpecifier).alias »«
-					ENDIF»«ENDFOR»«spacer»} from "«module»";'''
-			}
-		} else {
-			val importNode = findActualNodeFor(declaration);
-			return rewriteTokenText(importNode, spacer, SEMICOLON_INSERTED);
-		}
+		return resolutions
 	}
 
 	/**
-	 * Rewrites the node-content without comments.
-	 *
-	 * Inspired by {@link NodeModelUtils#getTokenText(INode)} but can treat any {@link LeafNodeWithSyntaxError syntax error}
-	 * nodes as a hidden one and reformats White-space after opening and before closing curly brace according to spacer-policy
-	 *
-	 * @param node the node to reformat.
-	 * @param spacer append after "{" and prepend before "}" can be empty string, then no white-space is inserted after
-	 * @param ignoredSyntaxErrorIssues - nodes of type LeafNodeWithSyntaxError having one of this issues are treated as ignored/hidden leafs
-	 * @returns modified textual form.
+	 * Obtains index based on the provided resource. Matches all broken names against object descriptions in the index.
+	 * Those that pass checks are added to the resolutions map as potential solutions.
+	 * <p>
+	 * Note that similar results can be achieved by querying scopes (i.e. TypeRef scope), as it was before. Unfortunately 
+	 * browsing the scope is much slower than index. For performance reasons we are using index, even though we need to 
+	 * duplicate some scope semantics, e.g. check for visibility. Still performance does not allow us to use scopes here.
 	 */
-	private static def String rewriteTokenText(ICompositeNode node, String spacer,
-		String... ignoredSyntaxErrorIssues) {
-		val StringBuilder builder = new StringBuilder(Math.max(node.getTotalLength(), 1));
-
-		var boolean hiddenSeen = false;
-		var boolean openingCurlySeen = false;
-		var boolean fixedASI = false;
-
-		for (ILeafNode leaf : node.getLeafNodes()) {
-			if(UtilN4.isIgnoredSyntaxErrorNode(leaf, SEMICOLON_INSERTED)){
-				fixedASI = true
-			}
-			if (!isHiddenOrIgnoredSyntaxError(leaf, ignoredSyntaxErrorIssues)) {
-				val text = leaf.getText();
-				if (builder.length() > 0) { // do not insert space before any content.
-					if (openingCurlySeen) {
-						// last real text was "{"
-						builder.append(spacer);
-					} else if ("}" == text) {
-						// this text is "}"
-						builder.append(spacer);
-					} else if ("," == text) {
-						// drop WS in front of commas.
-					} else {
-						// standard-WS handling.
-						if (hiddenSeen) {
-							builder.append(' ');
+	private def void addResolutionsFromIndex(Multimap<String, ImportableObject> resolution, IN4JSProject contextProject,
```

Comment:
Nice improvement! Looking up in the index rather than using scoping makes more sense to me.